### PR TITLE
Make conda’s package directories configurable

### PR DESCRIPTION
**Pull request by jezdez**
This adds the ability to use `D:` instead of `C:` on GitHub hosted Windows runners, to increase the speed of decompression.

See https://github.com/conda/conda-build/pull/5082#issuecomment-1845137430 for more information.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -637,6 +637,45 @@ jobs:
...
+          path: D:\conda_pkgs_dir
+          key:
+            ${{ runner.os }}-conda-${{ env.CACHE_NUMBER }}-${{
+            hashFiles('etc/example-environment.yml') }}
+      - uses: conda-incubator/setup-miniconda@v3
+        with:
+          activate-environment: anaconda-client-env
+          channel-priority: strict
+          environment-file: etc/example-environment.yml
+          pkgs_dirs: D:\conda_pkgs_dir
```
Comment (goanpeca) [inline README.md:669]: This should be `pkgs-dirs` :)

File: README.md [hunk 2]
```diff
@@ -637,6 +637,45 @@ jobs:
...
+```
+
+> Note:
+>
+> - GitHub hosted Windows runners are currently faster during cache
+>   decompression when configuring the packages directoies on the `D:` drive as
```
Comment (goanpeca) [inline README.md:675]: typo: directoies

**Comment by goanpeca** [comment]
Thanks for this work @jezdez !

Left a couple of minor comments 🚀

File: README.md (hunk 1 above)
Comment (jezdez) [reply README.md:669]: Good catch!

File: README.md [hunk 3]
```diff
@@ -637,6 +637,45 @@ jobs:
           use-only-tar-bz2: true # IMPORTANT: This needs to be set for caching to work properly!
 ```
 
+You may also set conda's packages directories (`CONDA_PKGS_DIRS`) config value,
+if you want to change setup-miniconda's default of `~/conda_pkgs_dir` with the
+`pkgs_dirs` config option. This is a comma-separated string like the channels
```
[the remark this answers is not in the document]
Comment (jezdez) [reply README.md:642]: Yup!

File: src/conda.ts [hunk 4]
```diff
@@ -130,9 +130,22 @@ export async function applyCondaConfiguration(
...
   }
 
+  // Package directories are also comma-separated, like channels
+  let pkgsDirs = utils.parsePkgsDirs(inputs.condaConfig.pkgs_dirs);
+  for (const pkgsDir of pkgsDirs) {
+    core.info(`Adding pkgs_dir '${pkgsDir}'`);
+    await condaCommand(["config", "--add", "pkgs_dirs", pkgsDir], options);
+  }
+  // We're also setting the appropriate conda config environment variable, to be safe
+  core.exportVariable("CONDA_PKGS_DIRS", pkgsDirs.join());
```
Comment (kenodegard) [inline src/conda.ts:140]:
```suggestion
  core.exportVariable("CONDA_PKGS_DIRS", pkgsDirs.join(","));
```

File: src/delete.ts [hunk 5]
```diff
@@ -5,34 +5,39 @@ import * as path from "path";
...
  * Clean up the conda cache directory
  */
 async function run(): Promise<void> {
   try {
-    const cacheFolder = utils.cacheFolder();
-
-    if (fs.existsSync(cacheFolder) && fs.lstatSync(cacheFolder).isDirectory()) {
+    const inputs = await core.group("Gathering Inputs...", input.parseInputs);
+    let pkgsDirs = utils.parsePkgsDirs(inputs.condaConfig.pkgs_dirs);
+    if (pkgsDirs.length) {
```
Comment (kenodegard) [inline src/delete.ts:18]: Can we invert this and exit early instead of deeply nesting everything?

```suggestion
    if (!pkgsDirs.length) return;
```

**Comment by jezdez** [comment]
@conda-incubator/setup-miniconda any takers for another review?

**Comment by jaimergp** [comment]
Code looks good!

Are there any issues with cross-drive hardlinking? I think Windows doesn't like creating those junctions over different units (or if the underlying drive is different), so maybe this is ends up becoming a performance hit _if_ the env is not in the same drive.

So, I wonder if:

- Should we also allow setting `envs-dirs` so users can request that new environments are created in the same drive?
- What about the default path for the installation when conda is bootstrapped from Miniconda or Miniforge installers? Should that also be configurable so we can have _everything_ in the same drive?

**Comment by jezdez** [comment]
I don't know why npm is adding all that cruft to dist/delete/index.js :(

**Comment by dbast** [comment]
> I don't know why npm is adding all that cruft to dist/delete/index.js :(

Afaik the reason is: every new function used from a dependency is copied into the index.js (`npm run build` = ncc = Compile a Node.js project into a single file)

**Review by jaimergp — APPROVED** [review]

**Pull request merged**